beacon-chain/rpc/eth/beacon/blocks.go

```diff
@@ -243,12 +243,104 @@ func (bs *Server) GetBlockSSZ(ctx context.Context, req *ethpbv1.BlockRequest) (*
 	return &ethpbv1.BlockSSZResponse{Data: sszBlock}, nil
 }
 
-func (bs *Server) GetBlockV2(ctx context.Context, request *ethpbv2.BlockRequestV2) (*ethpbv2.BlockResponseV2, error) {
-	panic("implement me")
+// GetBlockV2 retrieves block details for given block ID.
+func (bs *Server) GetBlockV2(ctx context.Context, req *ethpbv2.BlockRequestV2) (*ethpbv2.BlockResponseV2, error) {
+	ctx, span := trace.StartSpan(ctx, "beacon.GetBlockAltair")
+	defer span.End()
+
+	blk, err := bs.blockFromBlockID(ctx, req.BlockId)
+	if invalidBlockIdErr, ok := err.(*blockIdParseError); ok {
+		return nil, status.Errorf(codes.InvalidArgument, "Invalid block ID: %v", invalidBlockIdErr)
+	}
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get block from block ID: %v", err)
+	}
+	if blk == nil || blk.IsNil() {
+		return nil, status.Errorf(codes.Internal, "Could not find requested block")
+	}
+	phase0Blk, err := blk.PbPhase0Block()
+	// Assume we have an Altair block when Phase 0 block is unsupported.
+	if err != nil && !errors.Is(err, wrapper.ErrUnsupportedPhase0Block) {
+		return nil, status.Errorf(codes.Internal, "Could not check for phase 0 block")
+	}
+
+	if phase0Blk != nil {
+		v1Blk, err := migration.SignedBeaconBlock(blk)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not get signed beacon block: %v", err)
+		}
+		return &ethpbv2.BlockResponseV2{
+			Data: &ethpbv2.BeaconBlockContainerV2{
+				Block:     &ethpbv2.BeaconBlockContainerV2_Phase0Block{Phase0Block: v1Blk.Block},
+				Signature: v1Blk.Signature,
+			},
+		}, nil
+	}
+	altairBlk, err := blk.PbAltairBlock()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not check for Altair block")
+	}
+	v2Blk, err := migration.V1Alpha1BeaconBlockAltairToV2(altairBlk.Block)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get signed beacon block: %v", err)
+	}
+	return &ethpbv2.BlockResponseV2{
+		Data: &ethpbv2.BeaconBlockContainerV2{
+			Block:     &ethpbv2.BeaconBlockContainerV2_AltairBlock{AltairBlock: v2Blk},
+			Signature: blk.Signature(),
+		},
+	}, nil
 }
 
-func (bs *Server) GetBlockSSZV2(ctx context.Context, request *ethpbv2.BlockRequestV2) (*ethpbv2.BlockSSZResponseV2, error) {
-	panic("implement me")
+// GetBlockSSZV2 returns the SSZ-serialized version of the beacon block for given block ID.
+func (bs *Server) GetBlockSSZV2(ctx context.Context, req *ethpbv2.BlockRequestV2) (*ethpbv2.BlockSSZResponseV2, error) {
+	ctx, span := trace.StartSpan(ctx, "beacon.GetBlockSSZV2")
```

Comment:
This function is mostly duplicated, let's refactor

Reply:
+1, the only difference from this and GetBlockV2 is the return format. If you refactor to a common method that returns `*ethpb.SignedBeaconBlockAltair`, you would save 50 lines of code or so

Reply:
It's not that simple because I also have a `return` statement in the middle, inside `if phase0Blk != nil`. I refactored part of the function and reused the extracted code in all four state fetching endpoints.